src/d/d_pad.cpp

```diff
@@ -0,0 +1,894 @@
+#include "d/lyt/meter/d_lyt_meter.h"
+#include "d/snd/d_snd_player_mgr.h"
+#include "d/snd/d_snd_small_effect_mgr.h"
+#include "d/snd/d_snd_wzsound.h"
+#include "egg/core/eggController.h"
+#include "egg/core/eggHeap.h"
+#include "egg/math/eggMath.h"
+#include "egg/math/eggVector.h"
+#include "m/m_angle.h"
+#include "m/m_mtx.h"
+#include "m/m_pad.h"
+#include "m/m_vec.h"
+#include "nw4r/math/math_types.h"
+#include "rvl/KPAD/KPAD.h"
+#include "rvl/MTX/mtx.h"
+#include "rvl/MTX/mtxvec.h"
+#include "rvl/MTX/vec.h"
+#include "rvl/WPAD/WPAD.h"
+
+namespace dPad {
+
+ex_c ex_c::m_ex[4];
+ex_c *ex_c::m_current_ex = ex_c::m_ex;
+
+WPADInfo ex_c::m_info[2][4];
+
+bool ex_c::m_connected[4];
+
+void control_mpls_callback(s32 idx, s32 code) {
+    switch (code) {
+        case WPAD_ERR_OK: {
+            ex_c::on_0x54(idx);
+            break;
+        }
+        case WPAD_ERR_NO_CONTROLLER: {
+            ex_c::on_0x55(idx);
+            break;
+        }
+        case WPAD_ERR_BUSY: {
+            ex_c::on_0x56(idx);
+            break;
+        }
+        case 1: {
+            ex_c::on_0x57(idx);
+            break;
+        }
+    }
+}
+
+bool checkDeviceType(u32 type) {
+    switch (type) {
+        case WPAD_DEV_MOTION_PLUS:
+        case WPAD_DEV_MPLS_PT_FS:
+        case WPAD_DEV_MPLS_PT_CLASSIC:
+        case WPAD_DEV_MPLS_PT_UNKNOWN: return true;
+        default:                       return false;
+    }
+}
+
+void initMpls(s32 chan) {
+    KPADResetMpls(chan);
+    KPADDisableMplsZeroPlay(chan);
+    KPADSetMplsZeroPlayParam(chan, 0.005f);
+    KPADSetMplsZeroDriftMode(chan, 2);
+    disableMplsDirRevise(chan);
+    KPADSetMplsDirReviseParam(chan, 0.35f);
+    KPADEnableMplsAccRevise(chan);
+    KPADSetMplsAccReviseParam(chan, 0.03f, 0.4f);
+    KPADEnableMplsDpdRevise(chan);
+    KPADSetMplsDpdReviseParam(chan, 0.5f);
+}
+
+extern "C" void fn_80194080();
+void create() {
+    mPad::create();
+
+    KPADSetBtnRepeat(mPad::getCore(0)->getChannelID(), 0.5f, 0.1f);
+    ex_c::resetInfo();
+
+    KPADSetMplsWorkarea(new (0x20) u8[KPADGetMplsWorkSize()]);
+
+    for (int i = 0; i < 4; ++i) {
+        ex_c::m_ex[i].mMotion.init();
+        KPADSetReviseMode(i, 1);
+        KPADSetControlMplsCallback(i, control_mpls_callback);
+        ex_c::m_ex[i].resetState(i);
+    }
+    fn_80194080();
+}
+
+void setMpls(bool enable, s32 chan) {
+    if (enable) {
+        if (ex_c::m_ex[chan].mMplsEnabled == false) {
+            ex_c::m_ex[chan].mMplsEnabled = true;
+            KPADEnableMpls(chan, 5);
+            initMpls(chan);
+        }
+    } else {
+        if (ex_c::m_ex[chan].mMplsEnabled) {
+            ex_c::m_ex[chan].mMplsEnabled = false;
+            KPADDisableMpls(chan);
+        }
+    }
+}
+
+void centerPos(mVec2_c &in, mVec2_c &out) {
+    out.x = dGfx_c::getWidth4x3F() * 0.5f * (1.f + in.x) + dGfx_c::getWidth4x3LeftF();
+    out.y = dGfx_c::getCurrentScreenHeightF() * -0.5f * (1.f + in.y) + dGfx_c::getCurrentScreenTopF();
+}
```

Comment:
This looks like it converts Dpd coordinates `[-1f; 1f]^2` into screen coordinates `[-304; 304]x[-228; 228]`; accordingly `field_0x8` and `getDpdStableMaybe` store / return coordinates in screen space.